src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Extensibility/ProjectExportProvider.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Extensibility
+{
+    /// <summary>
+    /// MEF component which has methods for consumers to get to project specific MEF exports
+    /// </summary>
+    [Export(typeof(IProjectExportProvider))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class ProjectExportProvider : IProjectExportProvider
+    {
+        /// <summary>
+        /// This function requires going to the UI thread to map the hierarchy to the correct project
+        /// instance. 
+        /// </summary>
+        public T GetExport<T>(IVsHierarchy projectHierarchy) where T : class
+        {
+            T exportedValue = null;
+
+            if (projectHierarchy == null)
+            {
+                return exportedValue;
+            }
+            try
+            {
+                // The following code must be run on the UI thread
+                Shell.ThreadHelper.JoinableTaskFactory.Run(async () =>
```

Comment:
isn't it switching on UI thread on line 35?

---
In reply to: [71806554](https://github.com/dotnet/roslyn-project-system/pull/309#discussion_r71806554) [](ancestors = 71806554)